### PR TITLE
Refactor worker support

### PR DESCRIPTION
**Pull request by johnynek**
If we want to make it easier to add more worker implementations, I wanted to share our existing worker code (particularly for thrift support for workers).

This just does the refactoring. It also removes the guava dependency which was really minor (but Worker still uses it, I think, so pretty weak win).

@ianoc ?

### CODE REVIEW

File: src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+   * This is the reporter field for scalac, which we want to access
+   */
+  private static Field reporterField;
+  static {
+    try {
+      reporterField = Driver.class.getDeclaredField("reporter"); //NoSuchFieldException
+      reporterField.setAccessible(true);
+    }
+    catch (Exception ex) {
+      throw new RuntimeException("nope", ex);
```
Comment (softprops) [inline src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java]: I know this didn't change much before but I wonder if there isn't a better error message that can be reported

**Comment by ianoc** [comment]
LGTM

**Comment by johnynek** [comment]
@softprops what do you think of the latest?

**Comment by softprops** [comment]
Much better :) lgtm

**Pull request merged**